source/common/tls/ssl_socket.cc

```diff
@@ -241,6 +245,23 @@ void SslSocket::drainErrorQueue() {
     return;
   }
 
+  // Append detailed error info for certificate-related failures.
+  if (saw_cert_verify_failed || saw_no_client_cert) {
+    auto* extended_socket_info = reinterpret_cast<Envoy::Ssl::SslExtendedSocketInfo*>(
+        SSL_get_ex_data(rawSsl(), ContextImpl::sslExtendedSocketInfoIndex()));
+    if (extended_socket_info != nullptr) {
+      absl::string_view cert_validation_error = extended_socket_info->certificateValidationError();
+      if (!cert_validation_error.empty()) {
+        absl::StrAppend(&failure_reason_, ":", cert_validation_error);
+      } else if (saw_no_client_cert) {
+        // Provide a default message if no detailed error was captured
+        absl::StrAppend(&failure_reason_, ":peer did not provide required client certificate");
+      }
+    } else if (saw_no_client_cert) {
+      absl::StrAppend(&failure_reason_, ":peer did not provide required client certificate");
```

Comment:
Please refactor the logic so there's only one case that appends this error. Maybe something roughly like this:

```
bool added_reason = false;
if (saw_cert_verify_failed) {
  // checks for null
  StrAppend(failure_reason_, "the text");
  added_reason = true;
}
if (!added_reason && saw_no_client_cert) {
  StrAppend(......);
}
```